Received: (at 178209) by bugs.debian.org; 6 Mar 2003 23:41:35 +0000
From ecki@lina.inka.de Thu Mar 06 17:41:35 2003
Return-path: <ecki@lina.inka.de>
Received: from quechua.inka.de (mail.inka.de) [193.197.184.2] (mail)
	by master.debian.org with esmtp (Exim 3.12 1 (Debian))
	id 18r4zW-0005lZ-00; Thu, 06 Mar 2003 17:41:34 -0600
Received: from calista.inka.de (p508b2614.dip.t-dialin.net [80.139.38.20])
	by mail.inka.de with asmtp 
	id 18r4zU-0003YP-00; Fri, 07 Mar 2003 00:41:32 +0100
Received: from ecki by calista.inka.de with local   (Exim 3.35 #1 (Debian))
	id 18r4zS-0007tr-00; Fri, 07 Mar 2003 00:41:30 +0100
Date: Fri, 7 Mar 2003 00:41:30 +0100
To: Matt Domsch <Matt_Domsch@dell.com>
Cc: 178209@bugs.debian.org
Subject: Re: nameif segfault in net-tools-1.60
Message-ID: <20030306234130.GB29359@lina.inka.de>
References: <1046982565.1712.30.camel@localhost.localdomain> <20030306230708.GA29359@lina.inka.de> <1046992495.4526.19.camel@localhost.localdomain>
Mime-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Disposition: inline
In-Reply-To: <1046992495.4526.19.camel@localhost.localdomain>
User-Agent: Mutt/1.4i
From: Bernd Eckenfels <ecki@lina.inka.de>
Delivered-To: 178209@bugs.debian.org
X-Spam-Status: No, hits=-4.2 required=4.0
	tests=IN_REP_TO,PATCH_UNIFIED_DIFF,QUOTED_EMAIL_TEXT,REFERENCES,
	      SIGNATURE_SHORT_DENSE,SPAM_PHRASE_00_01,USER_AGENT,
	      USER_AGENT_MUTT
	version=2.44
X-Spam-Level: 

On Thu, Mar 06, 2003 at 05:14:55PM -0600, Matt Domsch wrote:
> 85748, filed against the Phoebe public beta so you can see it.

I think the following will do it:

Index: nameif.c
===================================================================
RCS file: /cvs/net-tools/nameif.c,v
retrieving revision 1.1
retrieving revision 1.3
diff -u -r1.1 -r1.3
--- nameif.c	18 Oct 2000 17:26:29 -0000	1.1
+++ nameif.c	6 Mar 2003 23:26:52 -0000	1.3
@@ -3,7 +3,7 @@
  * Writen 2000 by Andi Kleen.
  * Subject to the Gnu Public License, version 2.  
  * TODO: make it support token ring etc.
- * $Id: net-tools-1.60-cleanup-list-handling.patch,v 1.1 2004/03/29 07:34:04 seemant Exp $
+ * $Id: net-tools-1.60-cleanup-list-handling.patch,v 1.1 2004/03/29 07:34:04 seemant Exp $
  */ 
 #ifndef _GNU_SOURCE 
 #define _GNU_SOURCE
@@ -117,7 +117,8 @@
 }
 
 struct change { 
-	struct change *next,**pprev;
+	struct change *next;
+	int found;
 	char ifname[IFNAMSIZ+1];
 	unsigned char mac[6];
 }; 
@@ -139,10 +140,7 @@
 			ch->ifname, pos); 
 	if (parsemac(p,ch->mac) < 0) 
 		complain(_("cannot parse MAC `%s' at %s"), p, pos); 
-	if (clist) 
-		clist->pprev = &ch->next;
 	ch->next = clist;
-	ch->pprev = &clist;
 	clist = ch;
 	return 0; 
 }
@@ -200,7 +198,7 @@
 
 void usage(void)
 {
-	fprintf(stderr, _("usage: nameif [-c configurationfile] [-s] {ifname macaddress}")); 
+	fprintf(stderr, _("usage: nameif [-c configurationfile] [-s] {ifname macaddress}\n")); 
 	exit(1); 
 }
 
@@ -277,21 +275,21 @@
 		ch = lookupmac(mac); 
 		if (!ch) 
 			continue;
-			
-		*ch->pprev = ch->next;
+		
+		ch->found = 1;	
 		if (strcmp(p, ch->ifname)) { 
 			if (setname(p, ch->ifname) < 0)  
 				complain(_("cannot change name of %s to %s: %s"),
 						p, ch->ifname, strerror(errno)); 
 		} 
-		free(ch);
 	} 
 	fclose(ifh); 
 	
 	while (clist) { 
 		struct change *ch = clist;
 		clist = clist->next;
-		warning(_("interface '%s' not found"), ch->ifname); 
+		if (!ch->found)
+			warning(_("interface '%s' not found"), ch->ifname); 
 		free(ch); 
 	}
 




